packages/react/emotion-utils/src/createStyledComponent.tsx

```diff
@@ -0,0 +1,23 @@
+import { StyledComponent } from '@emotion/styled';
+import { AllHTMLAttributes } from 'react';
+
+type StyledBaseType<P> = StyledComponent<{ tag?: keyof JSX.IntrinsicElements } & P & AllHTMLAttributes<HTMLElement>>;
+
+export function createStyledComponent<P>(
+  StyledBase: StyledBaseType<P>
+): ({ ...rest }: React.ComponentProps<StyledBaseType<P>>) => React.ReactElement;
+export function createStyledComponent<P>(
+  StyledBase: StyledBaseType<P>,
+  tag: keyof JSX.IntrinsicElements
+): {
+  [tag in keyof JSX.IntrinsicElements]: ({ ...rest }: React.ComponentProps<StyledBaseType<P>>) => React.ReactElement;
+};
+export function createStyledComponent<P>(StyledBase: StyledBaseType<P>, tag?: keyof JSX.IntrinsicElements) {
+  if (!tag) {
+    return ({ ...rest }: React.ComponentProps<typeof StyledBase>) => <StyledBase {...rest} />;
+  } else {
+    return {
+      [tag]: ({ ...rest }: React.ComponentProps<typeof StyledBase>) => <StyledBase tag={tag} {...rest} />,
+    };
+  }
+}
```

Comment:
`StyledBaseType` doesn't guarantee that the returned element from the component is actually created as `tag`, for exmaple:

```tsx
const Component = styled((props: StyledBaseType<{}>) => {
  return createElement('div', props);
})``;

const ULComponent = createStyledComponent(Component, 'ul'); // but ULComponent is actually div
```